src/modules/ekf2/EKF/aid_sources/external_vision/ev_vel.h

```diff
@@ -0,0 +1,188 @@
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file ev_vel.h
+ * Handling different types of External Vision velocity measurements
+ */
+
+#ifndef ExternalVisionVel_H
+#define ExternalVisionVel_H
+
+#include "ekf.h"
+#include <uORB/topics/estimator_aid_source3d.h>
+
+class ExternalVisionVel
+{
+public:
+	ExternalVisionVel(Ekf &ekf_instance, const extVisionSample &vision_sample, const imuSample &imu_sample)
+		: ekf(ekf_instance), sample(vision_sample)
+	{
+
+		const Vector3f angular_velocity = imu_sample.delta_ang / imu_sample.delta_ang_dt - ekf._state.gyro_bias;
+		position_offset_body = ekf._params.ev_pos_body - ekf._params.imu_pos_body;
+		velocity_offset_body = angular_velocity % position_offset_body;
+	}
+
+	virtual ~ExternalVisionVel() = default;
+	virtual bool setMeasurement() = 0;
+	virtual bool fuseVelocity(estimator_aid_source3d_s &aid_src, float gate)
+	{
+		ekf.fuseLocalFrameVelocity(aid_src, aid_src.timestamp, measurement,
+					   measurement_var, gate);
+		return aid_src.fused;
+
+	}
+
+	virtual void resetVelocity()
+	{
+		ekf.resetVelocityTo(measurement, measurement_var);
+	}
+
+	void resetVerticalVelocity()
+	{
+		ekf.resetVerticalVelocityTo(measurement(2), measurement_var(2));
+	}
+
+	void enforceMinimumVariance()
+	{
+		min_variance = math::max(min_variance, sq(0.01f));
+
+		for (int i = 0; i < 3; ++i) {
+			measurement_var(i) = math::max(measurement_var(i), min_variance);
+		}
+	}
+
+	Ekf &ekf;
+	const extVisionSample &sample;
+	float min_variance;
+	Vector3f measurement;
+	Vector3f measurement_var;
+	Vector3f position_offset_body;
+	Vector3f velocity_offset_body;
+	Vector3f velocity_offset_earth;
+};
+
+class BodyFrameEV : public ExternalVisionVel
+{
+public:
+	BodyFrameEV(Ekf &ekf_instance, extVisionSample &vision_sample, const imuSample &imu_sample) :
+		ExternalVisionVel(ekf_instance, vision_sample, imu_sample) {}
+
+	void resetVelocity() override
+	{
+		const matrix::SquareMatrix<float, 3> rotated_variance = ekf._R_to_earth * matrix::diag(
+					measurement_var) * ekf._R_to_earth.transpose();
+		const Vector3f measurement_variance_ekf_frame = rotated_variance.diag() * 5.f; // variance bump
+		ekf.resetVelocityTo(ekf._R_to_earth * measurement, measurement_variance_ekf_frame);
+	}
+
+	void resetVerticalVelocity()
+	{
+		const matrix::SquareMatrix<float, 3> rotated_variance = ekf._R_to_earth * matrix::diag(
+					measurement_var) * ekf._R_to_earth.transpose();
+		const Vector3f measurement_variance_ekf_frame = rotated_variance.diag() * 5.f; // variance bump
+		ekf.resetVerticalVelocityTo((ekf._R_to_earth * measurement)(2, 0),
+					    measurement_variance_ekf_frame(2));
+	}
+
+	bool setMeasurement() override
```

Comment:
you have all the data in hand in the constructor so why not directly using it instead of having to save `velocity_offset_body` and wait for the `setMeasurement()` trigger? Otherwise you should rather send the measurement as an argument to that function: `setMeasurement(sample)`